File: Python/Sorting/MergeSort.py
def merge(a, low, mid, high):
    i = low
    j = mid+1
    b = []
    while i<= mid and j<= high:
        if a[i] < a[j]:
            x = a[i]
            b.append(x)
            i+=1
        else:
            x = a[j]
            b.append(x)
            j+=1
    while i<=mid:
        x = a[i]
        b.append(x)
        i+=1
    while j<= high :
        x = a[j]
        b.append(x)
        j+=1
    a[low:high+1] = b

def mergeSort(a, low, high):
    if low<high :
        mid = (low+high)//2

        mergeSort(a, low,mid)
        mergeSort(a, mid+1, high)
        merge(a, low, mid, high)

a = [65,3,276,87,5,5443,4,56,67,7687,21,45,6,78]
mergeSort(a, 0 , len(a)-1)
print(a)

File: C Programs/Class/mergeSort.c
//mergeSort

#include<stdio.h>
#define max 7

void merge(int a[], int low, int mid, int high){
	int i = low, j = mid+1, k = low;
	int b[max];
	
	while(i<=mid && j<=high){
		if(a[i] < a[j]){
			b[k] = a[i];
			i++; k++;
		}
		else{
			b[k] = a[j];
			j++; k++;
		}
	}
	while(i<=mid){
		b[k] = a[i];
		i++; k++;
	}
	while(j<=high){
		b[k] = a[j];
		j++; k++;
	}
	for(i=low;i<=high;i++){
		a[i] = b[i];
	}
}

void mergeSort(int a[], int low, int high){
	int mid;
	if(low<high){
		mid = (low+high)/2;
		
		mergeSort(a,low,mid);
		mergeSort(a, mid+1, high);
		
		merge(a,low,mid,high);
	}
}

int main(){
	int a[max] = {12,5,34,7,98,32,4};
	int i;
	
	printf("Before:\t");
	for(i=0;i<max;i++){
		printf("%d ", a[i]);
	}
	
	mergeSort(a, 0, max-1);
	
	printf("\nAfter:\t");
	for(i=0;i<max;i++){
		printf("%d ", a[i]);
	}
}
